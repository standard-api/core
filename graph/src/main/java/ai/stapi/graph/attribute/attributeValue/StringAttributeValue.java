package ai.stapi.graph.attribute.attributeValue;

public class StringAttributeValue extends AbstractAttributeValue<String> implements StringLikeAttributeValue<String> {

  public static final String SERIALIZATION_TYPE = "string";

  public StringAttributeValue(String value) {
    super(value.strip());
  }

  @Override
  public String getDataType() {
    return SERIALIZATION_TYPE;
  }

  @Override
  public String toStringValue() {
    return this.getValue();
  }
}
