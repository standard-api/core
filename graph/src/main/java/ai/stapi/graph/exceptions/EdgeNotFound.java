package ai.stapi.graph.exceptions;

import ai.stapi.identity.UniqueIdentifier;


public class EdgeNotFound extends GraphException {

  public EdgeNotFound(UniqueIdentifier edgeId, String edgeType) {
    super(
        String.format(
            "Edge with id \"%s\" of type \"%s\" not found.",
            edgeId.toString(),
            edgeType
        )
    );
  }

  public EdgeNotFound(UniqueIdentifier edgeId) {
    super(
        String.format(
            "Edge with id \"%s\" not found in any collection.",
            edgeId.toString()
        )
    );
  }
}
