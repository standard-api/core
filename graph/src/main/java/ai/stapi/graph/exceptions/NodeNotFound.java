package ai.stapi.graph.exceptions;

import ai.stapi.identity.UniqueIdentifier;


public class NodeNotFound extends GraphException {

  public NodeNotFound(UniqueIdentifier nodeId, String nodeType) {
    super(
        String.format(
            "Node with id \"%s\" of type \"%s\" not found.",
            nodeId.toString(),
            nodeType
        )
    );
  }

  public NodeNotFound(UniqueIdentifier nodeId) {
    super(
        String.format(
            "Node with id \"%s\" not found in any collection.",
            nodeId.toString()
        )
    );
  }
}
