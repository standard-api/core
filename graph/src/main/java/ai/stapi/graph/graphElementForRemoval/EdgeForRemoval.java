package ai.stapi.graph.graphElementForRemoval;

import ai.stapi.graph.traversableGraphElements.TraversableEdge;
import ai.stapi.identity.UniqueIdentifier;

public class EdgeForRemoval implements GraphElementForRemoval {

  public static final String ELEMENT_BASE_TYPE = "EDGE";
  private final UniqueIdentifier edgeId;
  private final String edgeType;

  public EdgeForRemoval(UniqueIdentifier edgeId, String edgeType) {
    this.edgeId = edgeId;
    this.edgeType = edgeType;
  }

  public EdgeForRemoval(TraversableEdge edge) {
    this.edgeId = edge.getId();
    this.edgeType = edge.getType();
  }

  @Override
  public UniqueIdentifier getGraphElementId() {
    return this.edgeId;
  }

  @Override
  public String getGraphElementType() {
    return this.edgeType;
  }

  @Override
  public String getGraphElementBaseType() {
    return ELEMENT_BASE_TYPE;
  }
}
