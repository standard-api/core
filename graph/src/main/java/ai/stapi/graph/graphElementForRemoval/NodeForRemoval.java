package ai.stapi.graph.graphElementForRemoval;

import ai.stapi.graph.graphelements.Node;
import ai.stapi.graph.traversableGraphElements.TraversableNode;
import ai.stapi.identity.UniqueIdentifier;

public class NodeForRemoval implements GraphElementForRemoval {

  public static final String ELEMENT_BASE_TYPE = "NODE";
  private final UniqueIdentifier nodeId;
  private final String nodeType;

  public NodeForRemoval(UniqueIdentifier nodeId, String nodeType) {
    this.nodeId = nodeId;
    this.nodeType = nodeType;
  }

  public NodeForRemoval(TraversableNode node) {
    this.nodeId = node.getId();
    this.nodeType = node.getType();
  }

  public NodeForRemoval(Node node) {
    this.nodeId = node.getId();
    this.nodeType = node.getType();
  }

  @Override
  public UniqueIdentifier getGraphElementId() {
    return this.nodeId;
  }

  @Override
  public String getGraphElementType() {
    return this.nodeType;
  }

  @Override
  public String getGraphElementBaseType() {
    return ELEMENT_BASE_TYPE;
  }
}
