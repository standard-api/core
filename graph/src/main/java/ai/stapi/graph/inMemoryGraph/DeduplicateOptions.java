package ai.stapi.graph.inMemoryGraph;

public enum DeduplicateOptions {
  SAME_EDGE_TYPES_BETWEEN_SAME_NODES
}
