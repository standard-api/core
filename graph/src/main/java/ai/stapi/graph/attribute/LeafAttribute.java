package ai.stapi.graph.attribute;


import ai.stapi.graph.attribute.attributeValue.AttributeValue;
import java.sql.Timestamp;
import java.util.Map;

public class LeafAttribute<T, AV extends AttributeValue<T>> extends AbstractAttribute<T> {

  public static final String DATA_STRUCTURE_TYPE = "LeafAttribute";
  private final AV boxedValue;

  public LeafAttribute(String name, Map<String, MetaData> metaData, AV boxedValue) {
    super(name, metaData);
    this.boxedValue = boxedValue;
  }

  public LeafAttribute(String name, Timestamp createdAt, AV boxedValue) {
    super(name, createdAt);
    this.boxedValue = boxedValue;
  }

  public LeafAttribute(String name, AV boxedValue) {
    super(name);
    this.boxedValue = boxedValue;
  }

  public AV getBoxedValue() {
    return boxedValue;
  }

  @Override
  public T getValue() {
    return this.getBoxedValue().getValue();
  }

  @Override
  public LeafAttribute<T, AV> getCopy() {
    return new LeafAttribute<>(this.getName(), this.getCreatedAt(), this.getBoxedValue());
  }
}
