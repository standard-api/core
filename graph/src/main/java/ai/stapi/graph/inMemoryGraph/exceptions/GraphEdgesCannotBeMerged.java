package ai.stapi.graph.inMemoryGraph.exceptions;

import ai.stapi.graph.exceptions.GraphException;
import ai.stapi.identity.UniqueIdentifier;
import java.util.List;
import org.apache.commons.lang3.StringUtils;

public class GraphEdgesCannotBeMerged extends GraphException {

  protected GraphEdgesCannotBeMerged(String message) {
    super("Graph edges could not be merged, " + message);
  }

  public static GraphEdgesCannotBeMerged becauseTheyHaveDifferentTypes() {
    return new GraphEdgesCannotBeMerged("because they have different types.");
  }

  public static GraphEdgesCannotBeMerged becauseTheyHaveDifferentIds() {
    return new GraphEdgesCannotBeMerged("because they have different ids.");
  }

  public static GraphEdgesCannotBeMerged becauseTheyHaveDifferentNodeIds() {
    return new GraphEdgesCannotBeMerged("because one or both nodes have different ids.");
  }

  public static GraphEdgesCannotBeMerged becauseThereIsMultipleEdgesGivenEdgeCouldBeMergeWith(
      List<UniqueIdentifier> possibleEdgesToMergeWith) {
    return new GraphEdgesCannotBeMerged(
        "There is multiple edges with which given edge could be merged with."
            + System.lineSeparator()
            + "Possible edges to be merged with: " +
            "[" + StringUtils.join(possibleEdgesToMergeWith, "], [") + "]."
    );
  }
}
