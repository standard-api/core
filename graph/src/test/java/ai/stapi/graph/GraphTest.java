package ai.stapi.graph;

import ai.stapi.graph.attribute.LeafAttribute;
import ai.stapi.graph.attribute.attributeValue.StringAttributeValue;
import ai.stapi.graph.exceptions.EdgeNotFound;
import ai.stapi.graph.exceptions.NodeNotFound;
import ai.stapi.graph.graphElementForRemoval.EdgeForRemoval;
import ai.stapi.graph.graphElementForRemoval.NodeForRemoval;
import ai.stapi.graph.inMemoryGraph.DeduplicateOptions;
import ai.stapi.graph.inMemoryGraph.exceptions.GraphEdgesCannotBeMerged;
import ai.stapi.graph.inputGraphElements.InputEdge;
import ai.stapi.graph.inputGraphElements.InputNode;
import ai.stapi.graph.test.base.UnitTestCase;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;

class GraphTest extends UnitTestCase {

  @Test
  void itShouldMergeTwoGraphsByIds() {
    var nodeFromInGraphG1 = new InputNode("merged_node_type");
    var nodeToInGraphG1 = new InputNode("node_type_to");
    var edgeE1InGraphG1 = new InputEdge(
        nodeFromInGraphG1,
        "edge_type",
        nodeToInGraphG1
    );

    var nodeFromInGraphG2 = new InputNode(
        nodeFromInGraphG1.getId(),
        nodeFromInGraphG1.getType()
    );
    var nodeToInGraphG2 = new InputNode(
        nodeToInGraphG1.getId(),
        nodeToInGraphG1.getType()
    );
    var edgeE1InGraphG2 = new InputEdge(
        edgeE1InGraphG1.getId(),
        nodeFromInGraphG2,
        edgeE1InGraphG1.getType(),
        nodeToInGraphG2
    );

    var originalAttribute =
        new LeafAttribute<>("original", new StringAttributeValue("original value"));
    var originalBAttribute =
        new LeafAttribute<>("original", new StringAttributeValue("original value"));
    var updatedOldAttribute = new LeafAttribute<>("updated", new StringAttributeValue("old value"));
    var updatedNewAttribute =
        new LeafAttribute<>("updated", new StringAttributeValue("updated value"));
    var newAttribute =
        new LeafAttribute<>("new", new StringAttributeValue("new value"));

    var graphG1 = new Graph(
        nodeFromInGraphG1.add(updatedOldAttribute).add(originalAttribute),
        nodeToInGraphG1,
        edgeE1InGraphG1.add(updatedOldAttribute).add(originalAttribute)
    );
    var graphG2 = new Graph(
        nodeFromInGraphG2.add(updatedNewAttribute).add(newAttribute).add(originalBAttribute),
        nodeToInGraphG2,
        edgeE1InGraphG2.add(updatedNewAttribute).add(newAttribute).add(originalBAttribute)
    );
    var actualGraph = graphG1.merge(graphG2);

    this.thenGraphApproved(actualGraph);
  }
  
//    
//    @Test
//    void itShouldNotAllowToReplaceNodeOfDifferentType()
//    {
//        var nodeA = new InputNode("type");
//        nodeA = nodeA.addIdentificator(new AttributeNodeIdentificator(
//            nodeA,
//            "identifying_attribute"
//        ));
//
//        var nodeB = new InputNode("different_type");
//        nodeB = nodeB.addIdentificator(new AttributeNodeIdentificator(
//            nodeB,
//            "identifying_attribute"
//        ));
//
//        nodeA = (InputNode) nodeA.add(new LeafAttribute<>("A_attribute", new StringAttributeValue("A_value")));
//        nodeB = (InputNode) nodeB.add(new LeafAttribute<>("B_attribute", new StringAttributeValue("B_value")));
//
//        var graph = new Graph(
//            nodeA
//        );
//
//        InputNode finalNodeA = nodeA;
//        InputNode finalNodeB = nodeB;
//        Executable executable = () -> graph.replaceNodeAndKeepEdges(
//            finalNodeA.getId(),
//            finalNodeB
//        );
//        Assertions.assertThrows(UnableToReplaceNode.class, executable);
//    }
//    
//    @Test
//    void itShouldReplaceNodes()
//    {
//        var nodeA = new InputNode("type");
//        nodeA = nodeA.addIdentificator(new AttributeNodeIdentificator(
//            nodeA,
//            "identifying_attribute"
//        ));
//
//        var nodeB = new InputNode("type");
//        nodeB = nodeB.addIdentificator(new AttributeNodeIdentificator(
//            nodeB,
//            "identifying_attribute"
//        ));
//
//        var nodeC = new InputNode("type");
//        nodeC = nodeC.addIdentificator(new AttributeNodeIdentificator(
//            nodeC,
//            "identifying_attribute"
//        ));
//
//        nodeA = (InputNode) nodeA.add(new LeafAttribute<>("A_attribute", new StringAttributeValue("A_value")));
//        nodeB = (InputNode) nodeB.add(new LeafAttribute<>("B_attribute", new StringAttributeValue("B_value")));
//        nodeC = (InputNode) nodeC.add(new LeafAttribute<>("irrelevant_attribute", new StringAttributeValue("irrelevant_value")));
//
//        var graph = new Graph(
//            nodeA,
//            nodeC
//        );
//
//        var actualGraph = graph.replaceNodeAndKeepEdges(nodeA.getId(), nodeB);
//
//        this.thenGraphApproved(
//            actualGraph
//        );
//    }

  @Test
  void itShouldRemoveEdge_AsGraphElementForRemoval() {
    var nodeFrom = new InputNode("node_from");
    var nodeTo = new InputNode("node_to");

    var alreadySavedEdge = new InputEdge(
        nodeFrom,
        "test_edge",
        nodeTo
    );
    alreadySavedEdge = alreadySavedEdge.addToEdge(
        new LeafAttribute<>("name", new StringAttributeValue("name")));
    var givenGraph = new Graph(nodeFrom, nodeTo, alreadySavedEdge);
    var edgeForRemoval = new EdgeForRemoval(alreadySavedEdge.getId(), alreadySavedEdge.getType());

    //When
    var actualGraph = givenGraph.removeGraphElement(edgeForRemoval);

    //Then
    InputEdge finalAlreadySavedEdge = alreadySavedEdge;
    Executable executable = () -> actualGraph.getEdge(
        finalAlreadySavedEdge.getId(),
        finalAlreadySavedEdge.getType()
    );
    Assertions.assertThrows(EdgeNotFound.class, executable);
  }

  @Test
  void itShouldRemoveNode_AsGraphElementForRemoval() {
    var alreadySavedNode = new InputNode("test_node");
    alreadySavedNode = alreadySavedNode.addToNode(
        new LeafAttribute<>("name", new StringAttributeValue("name")));

    var nodeForRemoval = new NodeForRemoval(alreadySavedNode.getId(), alreadySavedNode.getType());

    var givenGraph = new Graph(alreadySavedNode);
    //When
    var actualGraph = givenGraph.removeNode(nodeForRemoval);
    //Then
    InputNode finalAlreadySavedNode = alreadySavedNode;
    Executable executable = () -> actualGraph.getNode(finalAlreadySavedNode.getId(), "test_node");
    Assertions.assertThrows(NodeNotFound.class, executable);
  }

  @Test
  void itShouldMergeEdgesOfSameTypeBetweenSameNodes() {
    var originalGraph = new Graph();
    var firstOriginalNode = new InputNode("original_first_node");
    var secondOriginalNode = new InputNode("original_second_node");
    var firstOriginalEdge = new InputEdge(firstOriginalNode, "original_edge", secondOriginalNode);
    firstOriginalEdge = firstOriginalEdge.addToEdge(
        new LeafAttribute<>("original_attribute", new StringAttributeValue("anyValue")));
    originalGraph = originalGraph.withAll(firstOriginalNode, secondOriginalNode, firstOriginalEdge);

    var graphToBeMerged = new Graph();
    var edgeToBeMerged = new InputEdge(firstOriginalNode, "original_edge", secondOriginalNode);
    edgeToBeMerged = edgeToBeMerged.addToEdge(
        new LeafAttribute<>("merged_attribute", new StringAttributeValue("anyValue")));
    graphToBeMerged =
        graphToBeMerged.withAll(firstOriginalNode, secondOriginalNode, edgeToBeMerged);
    var actualGraph = originalGraph.traversable();
    actualGraph.merge(graphToBeMerged, DeduplicateOptions.SAME_EDGE_TYPES_BETWEEN_SAME_NODES);
    this.thenGraphApproved(actualGraph);
  }

  @Test
  void itThrowsErrorWhenEdgeCouldBeMergerWithMultipleEdges() {
    var originalGraph = new Graph();
    var firstOriginalNode = new InputNode("original_first_node");
    var secondOriginalNode = new InputNode("original_second_node");
    var firstOriginalEdge = new InputEdge(firstOriginalNode, "original_edge", secondOriginalNode);
    var secondOriginalEdge = new InputEdge(firstOriginalNode, "original_edge", secondOriginalNode);
    firstOriginalEdge = firstOriginalEdge.addToEdge(
        new LeafAttribute<>("original_attribute", new StringAttributeValue("anyValue")));
    secondOriginalEdge = secondOriginalEdge.addToEdge(
        new LeafAttribute<>("second_original_attribute", new StringAttributeValue("anyValue")));
    originalGraph = originalGraph.withAll(firstOriginalNode, secondOriginalNode, firstOriginalEdge,
        secondOriginalEdge);

    var graphToBeMerged = new Graph();
    var edgeToBeMerged = new InputEdge(firstOriginalNode, "original_edge", secondOriginalNode);
    edgeToBeMerged = edgeToBeMerged.addToEdge(
        new LeafAttribute<>("merged_attribute", new StringAttributeValue("anyValue")));
    graphToBeMerged =
        graphToBeMerged.withAll(firstOriginalNode, secondOriginalNode, edgeToBeMerged);
    Graph finalOriginalGraph = originalGraph;
    Graph finalGraphToBeMerged = graphToBeMerged;
    Executable throwable = () -> finalOriginalGraph
        .traversable()
        .merge(finalGraphToBeMerged, DeduplicateOptions.SAME_EDGE_TYPES_BETWEEN_SAME_NODES);
    Assertions.assertThrows(GraphEdgesCannotBeMerged.class, throwable);
  }
}
