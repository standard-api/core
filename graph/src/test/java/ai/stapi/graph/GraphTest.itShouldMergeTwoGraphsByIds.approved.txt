nodes:
	node_type: merged_node_type
	node_hash: 1A9CE
	node_edges:
		merged_node_type -> edge_type -> node_type_to (1A9CE -> 5BF34 -> 3F5B2)
	attributes:
		new -> new value
		original -> original value
		updated -> old value
		updated -> updated value

	node_type: node_type_to
	node_hash: 3F5B2
	node_edges:
		merged_node_type -> edge_type -> node_type_to (1A9CE -> 5BF34 -> 3F5B2)
	attributes:

edges:
	edge_type: edge_type
	edge_hash: 5BF34
	merged_node_type -> edge_type -> node_type_to (1A9CE -> 5BF34 -> 3F5B2)
	attributes:
		new -> new value
		original -> original value
		updated -> old value
		updated -> updated value
